cpp/src/arrow/compute/exec/exec_plan.cc

```diff
@@ -506,22 +506,49 @@ Result<ExecNode*> Declaration::AddToPlan(ExecPlan* plan,
   return node;
 }
 
+Declaration* Declaration::Root() {
+  Declaration* current = this;
+  while (!current->inputs.empty()) {
+    DCHECK_LE(current->inputs.size(), 1)
+        << "No clear root when a declaration has multiple inputs";
+    auto& input = current->inputs[0];
+    Declaration* maybe_next = input.get<Declaration>();
+    DCHECK(maybe_next) << "Attempt to get root when part of declaration is already built";
+    current = maybe_next;
+  }
+  return current;
+}
+
 Declaration Declaration::Sequence(std::vector<Declaration> decls) {
   DCHECK(!decls.empty());
 
   Declaration out = std::move(decls.back());
   decls.pop_back();
-  auto receiver = &out;
+  auto receiver = out.Root();
```

Comment:
Ah, fair enough.  I can switch to this model.  Thanks for the clarification.

Reply:
So this didn't end up working out because it dropped the tail.

plan = [b, c]
source = [a]
Sequence({source, plan.Root()) = [a, b]

I gave up on this and just switched to `std::vector<cp::Declaration>`